data-structures/disjoint-set/java/UnionFind.java

```diff
@@ -0,0 +1,82 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class UnionFind {
+    // Array indicating the root of the component each element is connected to.
+    protected int[] id;
+    // The index of the root of the component indicates the whole component's size.
+    protected int[] size;
+
+    public UnionFind(int n) {
+        id = new int[n];
+        size = new int[n];
+        // Initialize every element to be the root of its own component,
+        // and the size of each component to 1.
+        for (int i = 0; i < n; i++) {
+            id[i] = i;
+            size[i] = 1;
+        }
+    }
+
+    // Connects two components into a single one.
+    public void union(int p, int q) {
+        // Get the root of the components.
+        p = find(p);
+        q = find(q);
+
+        // If the root is the same, the elements are already connected.
+        if (p == q)
+            return;
+
+        // To keep the structure balanced with limited depth,
+        // add the smaller component as a branch of the larger component.
+        // Remember to increment the size of the resulting root.
+        if (size[p] > size[q]) {
+            // q's root is p, since p's component is bigger.
+            id[q] = p;
+            // p's component's size is incremented, since it now has all elements in the previous component.
+            size[p] += size[q];
+        } else {
+            id[p] = q;
+            size[q] += size[p];
+        }
+    }
+
+    // Finds the root of the component to which an element is connected.
+    public int find(int p) {
+        // Create a resizing array to compress the path of all elements along the way.
+        List<Integer> compressionList = new ArrayList<>();
```

Comment:
I would not rely on ArrayList, but rather another vector in the data structure. This ensures constant add time. Note that array list provides an amortized one.